proposals/2024-11-25_changing_wal_format.md

```diff
@@ -0,0 +1,257 @@
+  * when we at some point remove support of WAL v1, it's clear when it happens (v1 flag value is removed).
+* It allows users to set new version to write old format if needed (de-risking further).
+* Gives devs quicker feedback, makes testing easier, and motivates further contributions.
+
+We propose to document that behaviour in:
+
+* dedicated guide for users
+* flag help
+* tsdb/docs for devs
+
+To achieve WAL versioning we also propose to start versioning the WAL, as [a whole format](https://github.com/prometheus/prometheus/blob/main/tsdb/docs/format/wal.md). This is necessary to communicate a breaking change and to tell Prometheus what WAL format to write in.
+
+We propose the addition of a `meta.json` file in the wal directory, similar to [block meta.json](https://github.com/prometheus/prometheus/blob/5e124cf4f2b9467e4ae1c679840005e727efd599/tsdb/block.go#L171), with Version field set to `1` for the current format and `2` for new changes e.g. when we start to write [new records](https://github.com/prometheus/prometheus/pull/15467/files). No `meta.json` is equivalent to `{"version":1} `meta.json` file.
+
+We propose to also store the new WALs in separate directories e.g. `wal.v2`. Thanks to that the rewrite from one version to another is eventual and can be done segment by segment.
+The additional advantage of this way of versioning is that it's clear when your WAL fully migrated to a certain version.
+
+Finally, we propose that all Prometheus releases will contain the following table:
+
+| Data        | Supported | Writes |
+|-------------|-----------|--------|
+| WAL         | v1, v2    | v1     |
+| Block index | v2        | v2     |
+
+```
+The last revertable Prometheus version: v3.0 
+```
+
+See [alternatives](#alternatives) for other ideas.
+
+### Two-Fold Migration Strategy
+
+Given the following example:
+
+![twofold.png](../assets/2024-11-25_changing_wal_format/twofold.png)
+
+1. We release Prometheus X+1 version that supports both Y and Y+1 data but still writes Y.
+2. We release Prometheus X+2 version that supports both Y and Y+1 data, but now it writes new data as Y.
+
+While this example shows only one version where of forward compatibility (when Y and Y+1 are supported, but Y is still written), in practice there could be more "forward compatible releases" within this strategy.
+
+Pros:
+* Users have a durable rollout path back and forth.
+* Dev has clarity on how to develop "breaking revert" changes.
+* In theory, it allows revert to X from X+2, by going through X+1 and ensuring all data was migrated (eventually) to Y version. In practice however that eventuality is long, or hard to discover.
+
+Cons:
+* It can catch users by surprise
+  * **Mitigation**: We also plan guide and documentation.
+* It takes time to rollout new changes if you want them fast.
+  * **Mitigation**: We plan to add a flag to opt-in sooner
+* It is a breaking change which shouldn't be made in a major release.
```

Comment:
Did you mean minor?
Also, is changing the default write format really a breaking change? The user has to do nothing and they can downgrade. The breaking change comes when you remove support for version Y.

I think the way around deferring removal of the old version Y to major release is to introduce a deprecation policy. Set Y+1 as default, deprecate the old version and mark it in docs/startup message. Then remove after 2 minor versions or more.
This would only affect people that actually set Y as explicit version.

```suggestion
* It is a breaking change which shouldn't be made in a minor release.
```

Reply:
Yes I meant minor. 

Or we could try to keep it until the next major version WDYT? I think I would propose to not set it in stone - depends how easy is to maintain old format support.